core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -278,9 +325,38 @@ public void runWith(Substream substream) {
     }
 
     Substream substream = createSubstream(0);
+    if (isHedging) {
+      Precondition precondition = new Precondition() {
+        @Override
+        @GuardedBy("RetriableStream.this.lock")
+        public boolean met() {
+          return hasPotentialHedging() && (throttle == null || throttle.isAboveThreshold());
+        }
+      };
+      rescheduler = new ConcurrentRescheduler(
+          new HedgingRunnable(), scheduledExecutorService, precondition, lock);
+    }
+
     drain(substream);
+  }
 
-    // TODO(zdapeng): schedule hedging if needed
+  private final class HedgingRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      callExecutor.execute(new Runnable() {
+        @Override
+        public void run() {
+          int previousAttempts;
+          synchronized (lock) {
```

Comment:
I'm really surprised this doesn't check `hasPotentialHedging` or the precondition.